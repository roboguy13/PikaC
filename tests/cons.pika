%generate cons

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

cons : Int -> Sll -> Sll;
cons i Nil := Cons i (Nil [Sll]) [Sll];
cons i (Cons a as) := Cons i (Cons a as [Sll]) [Sll];