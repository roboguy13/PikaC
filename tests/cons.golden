void cons(loc i1, loc x, loc x2);
void cons(loc i1, loc x, loc x2) {
 if (x == 0) {
  loc x = (loc)malloc(2 * sizeof(loc));
  WRITE_LOC(x2, 0, x);
  WRITE_LOC(x, 0, i1);
  WRITE_LOC(x, 1, 0);
 } else {
  loc x4 = (loc)malloc(2 * sizeof(loc));
  WRITE_LOC(x2, 0, x4);
  WRITE_LOC(x4, 0, i1);
  WRITE_LOC(x4, 1, x);
 }
}